### PR TITLE
feat: move range filter to global header

### LINKED ISSUE #1815: Feature: Insights/Lists - move date filter from individual charts to header
### Type of feature

🍕 Feature

### Current behavior

Right now, on Insight Pages there is no global date filter. The only filterable component is the repositories and contributors table, making it a bit confusing.

For the initial design of the list feature we wanted to do individual filters for each chart, but realized a global one would be better.

## Goal
Implement global date filter on Lists and Insight Pages.
- Add the date filter to the header on every tab of the List and Insight pages
- It should be usable on every tab (we can disable it on the reports tab until we figure out what we want to do with reports)
- The state should persist as the user navigates the tabs. ie. If I'm on the Dashboard Tab and I select "30d", if I then navigate to the Contributors tab, the filter value should still be "90d"
- filter values should be 7d / 30d / 90d  (it's currently 3m instead of 90d on some screens)
- default value should be 30d

## Mockups
These are quick mockups so alignment might not be perfect. Just align the filter vertically to the navigation tabs, and horizontally it should align to the right with the rest of the components on the page.

Insight pages:
image

Lists:
image

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Contributing Docs

- [X] I agree to follow this project's Contribution Docs

### PR DESCRIPTION
**Pull request by OgDev-01**
## Description
This PR:

Introduces changes to improve range filter management by shifting from using useState to utilizing URL query parameters. This modification enhances the usability and maintainability of the code while providing benefits such as improved navigation and bookmarking of filter states.

### Benefits
- Enhanced Usability: Users can easily bookmark or share filter configurations, simplifying navigation.
- Better User Experience: Seamless synchronization between the filter and URL parameters provides a more intuitive experience.

## What type of PR is this? (check all applicable)

- [ ] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [x] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents
closes #1815 
related to https://github.com/open-sauced/app/issues/712

## Mobile & Desktop Screenshots/Recordings

https://github.com/open-sauced/app/assets/62995161/91d76e84-53dc-4867-a706-02fd07152193

## Added tests?

- [ ] 👍 yes
- [x] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 docs.opensauced.pizza
- [ ] 🍕 dev.to/opensauced
- [ ] 📕 storybook
- [x] 🙅 no documentation needed

### CODE REVIEW

File: components/organisms/Contributors/contributors.tsx [hunk 1]
```diff
@@ -24,11 +25,11 @@ interface ContributorProps {
 
 const Contributors = ({ repositories }: ContributorProps): JSX.Element => {
   const router = useRouter();
-  const topic = router.query.pageId as string;
```
Comment (brandonroberts) [inline components/organisms/Contributors/contributors.tsx:27]: nit: revert this change

**Review by brandonroberts — APPROVED** [review]

**Comment by OgDev-01** [comment]
Hey @nickytonline, can you take a peak at this when you have the time?

**Comment by OgDev-01** [comment]
This PR has a lot of files changed because of the change in range type cc: @nickytonline

File: components/molecules/ComponentDateFilter/component-date-filter.tsx [hunk 2]
```diff
@@ -22,15 +24,24 @@ const ComponentDateFilter = ({ setRangeFilter, defaultRange }: ComponentDateFilt
...
-        </div>
+          <div
+            onClick={() => handleFilterClick(range)}
+            className={`px-4 py-1.5 rounded-lg cursor-pointer transition text-light-slate-9 ${
+              activeFilter === range && "border text-light-slate-12 bg-light-slate-5"
+            }`}
+            key={index}
+          >
+            {rangeFormatter(range)}
+          </div>
```
Comment (nickytonline) [inline components/molecules/ComponentDateFilter/component-date-filter.tsx:43]: Refactor (blocking): this needs to be a button. It doesn't have to be our button component. It can be just a regular `<button />`. It will also need a label. You can wrap it with a label element and add a span inside for screen readers only, e.g. `<span class name="sr-only">some label text</span>`

```suggestion
          <button
            onClick={() => handleFilterClick(range)}
            className={`px-4 py-1.5 rounded-lg cursor-pointer transition text-light-slate-9 ${
              activeFilter === range && "border text-light-slate-12 bg-light-slate-5"
            }`}
            key={index}
          >
            {rangeFormatter(range)}
          </button>
```

File: components/organisms/Contributors/contributors.tsx [hunk 3]
```diff
@@ -13,11 +13,11 @@ import { calcDistanceFromToday } from "lib/utils/date-utils";
 import SkeletonWrapper from "components/atoms/SkeletonLoader/skeleton-wrapper";
 import LimitSelect from "components/atoms/Select/limit-select";
 
+import ClientOnly from "components/atoms/ClientOnly/client-only";
```
Comment (nickytonline) [inline components/organisms/Contributors/contributors.tsx:16]: This doesn't appear to be related to the PR. It looks like the import was moved.

File: components/organisms/Contributors/contributors.tsx [hunk 4]
```diff
@@ -183,31 +183,30 @@ const Contributors = ({ repositories }: ContributorProps): JSX.Element => {
...
+              <ContributorCard
+                key={index}
+                className=""
+                contributor={{ ...contributor }}
+                topic={topic}
+                repositories={repositories}
+              />
+            ))}
+          </div>
+        </ClientOnly>
```
Comment (nickytonline) [inline components/organisms/Contributors/contributors.tsx:206]: refactor (blocking): render as much markup server side that doesn't contribute to a hydration error.

```suggestion
        <div className="grid w-full gap-3 grid-cols-automobile md:grid-cols-autodesktop">
          <ClientOnly>
            {isLoading ? <SkeletonWrapper height={210} radius={12} count={9} /> : ""}
            {isError ? <>An error occurred!..</> : ""}
            {contributorArray.map((contributor, index) => (
              <ContributorCard
                key={index}
                className=""
                contributor={{ ...contributor }}
                topic={topic}
                repositories={repositories}
              />
            ))}
          </ClientOnly>
        </div>
```

**Review by nickytonline — CHANGES_REQUESTED** [review]
I'm just on the bus, so will do a more thorough review once I'm at a desk, but here's some initial feedback.

File: components/organisms/ToolList/nav-item.tsx [hunk 5]
```diff
@@ -13,6 +14,8 @@ interface NavItemProps {
 
 const NavItem: React.FC<NavItemProps> = ({ username, filterName, tool, selectedFilter, selectedTool }) => {
   const [total, setTotal] = useState<number>();
+  const router = useRouter();
+  const { range } = router.query;
```
Comment (nickytonline) [inline components/organisms/ToolList/nav-item.tsx:18]: refactor (non-blocking): Set a default value here and don't bother with the `?? "30"` below.

```suggestion
  const { range = "30" } = router.query;
```

File: components/organisms/ToolList/nav-item.tsx [hunk 6]
```diff
@@ -21,7 +24,11 @@ const NavItem: React.FC<NavItemProps> = ({ username, filterName, tool, selectedF
   return (
     <Link
       href={`/${username ? `pages/${username}/` : ""}${filterName}/${tool.name.toLowerCase()}${
-        selectedFilter ? `/filter/${Array.isArray(selectedFilter) ? selectedFilter.join("/") : selectedFilter}` : ""
+        selectedFilter
+          ? `/filter/${Array.isArray(selectedFilter) ? selectedFilter.join("/") : selectedFilter}?range=${
+              range ?? "30"
+            }`
+          : `?range=${range ?? "30"}`
```
Comment (nickytonline) [inline components/organisms/ToolList/nav-item.tsx:31]: refactor (non-blocking): This in conjunction with `const { range = "30" } = router.query` above.

```suggestion
          : `?range=${range}`
```

File: layouts/filter.tsx [hunk 7]
```diff
@@ -1,15 +1,25 @@
...
 const FilterLayout = ({ children }: { children: React.ReactNode }) => {
   const { toolList, selectedTool, filterName, selectedFilter, userOrg } = useNav();
+  const router = useRouter();
+  const { range } = router.query;
+
+  useEffect(() => {
+    if (!range) {
+      setQueryParams({ range: "30" });
+    }
+  }, [range, router.isReady]);
```
Comment (nickytonline) [inline layouts/filter.tsx:22]: Refactor (potentially blocking): This seems unnecessary as the default value of 30 would be used if none was present in the querystring.

```suggestion
```

File: layouts/hub-page.tsx [hunk 8]
```diff
@@ -11,15 +11,22 @@ import useNav from "lib/hooks/useNav";
...
   const insightId = pageId as string;
   const { data: insight, isLoading, isError } = useInsight(insightId);
   const repositories = insight?.repos.map((repo) => repo.repo_id);
 
+  useEffect(() => {
+    if (!range) {
+      setQueryParams({ range: "30" });
+    }
+  }, [range, router.isReady]);
+
```
Comment (nickytonline) [inline layouts/hub-page.tsx:29]: Same as above.

```suggestion

```

File: lib/hooks/api/useContributors.ts [hunk 9]
```diff
@@ -48,7 +48,7 @@ const useContributors = (intialLimit = 10, repoIds: number[] = [], range = 30) =
     query.delete("topic");
   }
 
-  query.set("range", `${range}`);
+  query.set("range", `${range ?? 30}`);
```
Comment (nickytonline) [inline lib/hooks/api/useContributors.ts:51]: nit: set the default value above

`const { pageId, selectedFilter, range = 30 } = router.query;`

File: pages/lists/[listId]/activity.tsx [hunk 10]
```diff
@@ -80,21 +83,33 @@ export const getServerSideProps = async (ctx: GetServerSidePropsContext) => {
...
     setContributorType,
     contributorType,
-  } = useMostActiveContributors({ listId: list!.id, initData: activityData.contributorStats.data, range });
+  } = useMostActiveContributors({ listId: list!.id, initData: activityData.contributorStats.data });
+
+  useEffect(() => {
+    if (!range) {
+      setQueryParams({ range: "30" });
+    }
+  }, [range]);
```
Comment (nickytonline) [inline pages/lists/[listId]/activity.tsx:101]: Same as above. No need to update the URL if using the default value.

```suggestion
```

File: pages/lists/[listId]/contributors.tsx [hunk 11]
```diff
@@ -54,17 +57,24 @@ interface ContributorListPageProps {
...
   // create useIsOwner(list?.user_id, userId) once we're ready to implement this.
   const isOwner = false;
+  const router = useRouter();
+  const { range } = router.query;
+
+  useEffect(() => {
+    if (!range) {
+      setQueryParams({ range: "30" });
+    }
+  }, [range]);
```
Comment (nickytonline) [inline pages/lists/[listId]/contributors.tsx:67]: Same as above. No need for the setting the range query param.

```suggestion
  const { range = "30" } = router.query;
```

**Review by nickytonline — CHANGES_REQUESTED** [review]
I've tested this locally and it works great! Just a few change requests and then this is good to go.

File: components/molecules/ComponentDateFilter/component-date-filter.tsx [hunk 12]
```diff
@@ -22,15 +24,27 @@ const ComponentDateFilter = ({ setRangeFilter, defaultRange }: ComponentDateFilt
...
+          ${rangeFormatter(range)}
+          ${range === 7 ? "from today" : "from the last commit"}
+        `}
+          direction="top"
         >
-          {rangeFormatter(range)}
-        </div>
+          <span className="sr-only">
+            {rangeFormatter(range)} {rangeFormatter(range)}${range === 7 ? "from today" : "from the last commit"}
+          </span>
```
Comment (nickytonline) [inline components/molecules/ComponentDateFilter/component-date-filter.tsx:37]: I made a mistake suggesting the label here. The button already has text. We can improve the accesibility in another PR to describe what the range does. I'll create a separate issue for that.

Comment (nickytonline) [reply components/molecules/ComponentDateFilter/component-date-filter.tsx:37]: Remove the label I suggested, and I think this is good to go @OgDev-01.

**Comment by OgDev-01** [comment]
Can you take a look now @nickytonline?

**Review by nickytonline — APPROVED** [review]
Thanks @OgDev-01! 🚀

File: components/molecules/ComponentDateFilter/component-date-filter.tsx (hunk 12 above)
Comment (nickytonline) [reply components/molecules/ComponentDateFilter/component-date-filter.tsx]: I created #212 for this.
